File: operonx/core/tracing/emitter.py
"""EventEmitter — what ops + scheduler call to record trace events.

All methods are sync + O(1) per the execution model in
``docs/TRACING_REDESIGN_PLAN.md`` §3.7. They append to the bound pipeline's
buffer and (for some kinds) bookkeep an in-flight start_time map for
``OP_END(status="cancelled")`` synthesis (Rule 3).

The emitter is bound per ``engine.start()`` call. ``NullEmitter`` is bound
when no tracer is configured — its methods are no-ops.

Two ContextVars carry the per-call binding through ``asyncio.create_task``:

- ``_current_emitter_var``: the active emitter (set once at engine.start)
- ``_current_op_var``: the ``(op_name, ctx)`` of the op currently executing,
  set/cleared by ``_pump`` around each invocation

Used together to scope ``annotate(key, value)`` to the current op without
threading inputs.
"""

from __future__ import annotations

from contextlib import contextmanager
from contextvars import ContextVar
from datetime import datetime, timezone
from time import perf_counter
from typing import TYPE_CHECKING, Any, Dict, Iterator, Optional, Tuple

from operonx.core.tracing.events import EventKind, TraceEvent

if TYPE_CHECKING:
    from operonx.core.tracing.pipeline import TracePipeline


_current_emitter_var: ContextVar["EventEmitter"] = ContextVar(
    "operonx_current_emitter"
)
"""Per-call binding of the active emitter. Set by ``engine.start()``."""

_current_op_var: ContextVar[Tuple[str, Tuple[str, ...]]] = ContextVar(
    "operonx_current_op"
)
"""``(op_name, ctx)`` of the op currently executing. Set/cleared by ``_pump``
around each op invocation."""


def current_emitter() -> "EventEmitter":
    """Return the emitter bound to this call, or ``NullEmitter`` if none.

    Use from inside an op when you need to call ``annotate`` or emit a
    custom event. Returning a singleton ``NullEmitter`` means callers never
    have to ``if emitter is None`` — invoking methods is always safe.
    """
    try:
        return _current_emitter_var.get()
    except LookupError:
        return _NULL_EMITTER


class EventEmitter:
    """Pushes events into the bound pipeline's buffer.

    Sync, O(1) per method. The execution model (§3.7) forbids any
    processor or exporter work on the emit path — that all happens inside
    ``pipeline.flush()`` which runs as a separate ``asyncio.create_task``.
    """

    def __init__(self, pipeline: "TracePipeline", request_id: str) -> None:
        self._pipeline = pipeline
        self._request_id = request_id
        self._seq = 0
        self._start_times: Dict[Tuple[str, Tuple[str, ...]], float] = {}
        """``(op_name, ctx) → perf_counter()`` at OP_START. Looked up by
        ``start_time_of`` for cancel-emit duration_ms (Rule 3)."""

    # ------------------------------------------------------------------
    # Core emit
    # ------------------------------------------------------------------

    def emit(self, event: TraceEvent) -> None:
        """Append an event to the pipeline buffer.

        Called by helpers below or directly by user code that constructs
        bespoke events (rare).
        """
        self._pipeline._push(event)

    # ------------------------------------------------------------------
    # Convenience helpers — construct + emit the common kinds
    # ------------------------------------------------------------------

    def op_start(
        self,
        op_name: str,
        ctx: Tuple[str, ...],
        inputs: Optional[Dict[str, Any]] = None,
        media_refs: Optional[list] = None,
    ) -> None:
        """Fired in ``_pump`` after ``get_inputs()`` resolves, before
        ``_exec_core()`` enters.

        ``media_refs`` is a parallel ``list[MediaRef]`` produced by
        ``extract_media`` over the trace-time view of inputs. Bytes live on
        the refs (not in event memory beyond what they'd cost in the I/O
        dict anyway); the exporter uploads them at flush and substitutes
        Langfuse media tokens for the ``<media:N>`` placeholders left in
        ``inputs``. Empty list when no Media instances were found.
        """
        self._start_times[(op_name, ctx)] = perf_counter()
        self.emit(self._build(EventKind.OP_START, op_name, ctx, {
            "inputs": inputs or {},
            "media_refs": media_refs or [],
        }))

    def op_end(
        self,
        op_name: str,
        ctx: Tuple[str, ...],
        outputs: Optional[Dict[str, Any]] = None,
        status: str = "ok",
        duration_ms: Optional[float] = None,
        yield_count: int = 0,
        media_refs: Optional[list] = None,
    ) -> None:
        """Fired in ``_pump``'s ``finally`` after ``store_result``.

        ``status`` is one of ``"ok"``, ``"error"``, ``"cancelled"``.
        ``duration_ms`` defaults to elapsed since OP_START. Pops the in-flight
        entry — calling twice for the same op is a no-op the second time
        (idempotent), needed for the cancel-double-emit guard in Rule 3.

        ``media_refs`` mirrors ``op_start``'s parameter, for outputs.
        """
        key = (op_name, ctx)
        if key not in self._start_times:
            # Already ended (cancel-emit raced with finally) — drop second.
            return
        if duration_ms is None:
            duration_ms = (perf_counter() - self._start_times[key]) * 1000.0
        del self._start_times[key]
        self.emit(self._build(EventKind.OP_END, op_name, ctx, {
            "outputs": outputs or {},
            "status": status,
            "duration_ms": duration_ms,
            "yield_count": yield_count,
            "media_refs": media_refs or [],
        }))

    def op_yield(
        self,
        op_name: str,
        ctx: Tuple[str, ...],
        yielded: Dict[str, Any],
        idx: int,
        media_refs: Optional[list] = None,
    ) -> None:
        """Fired per-yield from a generator op, gated by ``emit_yields=N``.

        ``media_refs`` carries any Media stripped out of ``yielded`` so the
        event buffer doesn't hold raw bytes. Currently unused by the default
        Langfuse exporter (which folds yields into the parent observation's
        ``last_yielded`` metadata) — present for forward-compat with custom
        exporters that render per-yield observations.
        """
        self.emit(self._build(EventKind.OP_YIELD, op_name, ctx, {
            "yielded": yielded,
            "idx": idx,
            "media_refs": media_refs or [],
        }))

    def annotate(self, key: str, value: Any) -> None:
        """Attach metadata to the currently executing op.

        Reads ``_current_op_var`` (set by ``_pump``). Raises ``LookupError``
        if called outside an op scope — programming error.
        """
        op_name, ctx = _current_op_var.get()
        self.emit(self._build(EventKind.ANNOTATION, op_name, ctx, {
            "key": key,
            "value": value,
        }))

    def llm_usage(
        self,
        op_name: str,
        ctx: Tuple[str, ...],
        model: str,
        prompt_tokens: int = 0,
        completion_tokens: int = 0,
        total_tokens: int = 0,
        cost_usd: float = 0.0,
    ) -> None:
        """Token / cost report. MUST be called before the matching OP_END."""
        self.emit(self._build(EventKind.LLM_USAGE, op_name, ctx, {
            "model": model,
            "prompt_tokens": prompt_tokens,
            "completion_tokens": completion_tokens,
            "total_tokens": total_tokens,
            "cost_usd": cost_usd,
        }))

    def media_ref(
        self,
        op_name: str,
        ctx: Tuple[str, ...],
        handle: str,
        mime: str,
        size_bytes: int,
    ) -> None:
        """Reference a blob stored separately in the pipeline's MediaStore.
        Bytes are NOT in the event payload — call
        ``pipeline.media_store.put(handle, data, mime)`` first."""
        self.emit(self._build(EventKind.MEDIA_REF, op_name, ctx, {
            "handle": handle,
            "mime": mime,
            "size_bytes": size_bytes,
        }))

    @contextmanager
    def group(self, name: str, **metadata: Any) -> Iterator[None]:
        """Manual logical group via ``with emitter.group("xyz"): ...``.

        Most users won't call this directly — ``GroupBy`` processor emits
        ``GROUP_START``/``GROUP_END`` based on event boundaries in the
        stream. Use this when grouping is structural in user code rather
        than derived from events.
        """
        self.emit(self._build(EventKind.GROUP_START, None, (), {
            "name": name, **metadata,
        }))
        try:
            yield
        finally:
            self.emit(self._build(EventKind.GROUP_END, None, (), {
                "name": name, "status": "ok",
            }))

    # ------------------------------------------------------------------
    # Read-only accessors used by the scheduler for cancel-emit (Rule 3)
    # ------------------------------------------------------------------

    def start_time_of(
        self,
        op_name: str,
        ctx: Tuple[str, ...],
    ) -> float:
        """``perf_counter()`` snapshot from this op's OP_START, or 0.0 if
        unknown (e.g. cancel-before-start, where OP_START never fired)."""
        return self._start_times.get((op_name, ctx), 0.0)

    # ------------------------------------------------------------------
    # Internal
    # ------------------------------------------------------------------

    def _build(
        self,
        kind: EventKind,
        op_name: Optional[str],
        ctx: Tuple[str, ...],
        payload: Dict[str, Any],
    ) -> TraceEvent:
        """Construct a TraceEvent with this emitter's seq + request_id.

        ``event_id`` is ``"{request_id}-{seq}"`` rather than UUID4 — uniqueness
        within a pipeline is the only requirement, and avoiding ``uuid.uuid4()``
        keeps the emit hot path under budget (§3.7 TR13).
        """
        seq = self._seq
        self._seq = seq + 1
        return TraceEvent(
            event_id=f"{self._request_id}-{seq}",
            request_id=self._request_id,
            kind=kind,
            op_name=op_name,
            ctx=ctx,
            timestamp=datetime.now(timezone.utc),
            seq=seq,
            payload=payload,
        )


class NullEmitter(EventEmitter):
    """Bound when no tracer is configured. All methods are no-ops.

    The fast path: op code calls ``emit_*`` blindly; cost is one method-table
    dispatch (~ns). Subclasses ``EventEmitter`` so type checks against the
    base class still work.
    """

    def __init__(self) -> None:
        # Do NOT call super().__init__ — NullEmitter never needs a pipeline.
        # Skipping it avoids allocating the start_times dict.
        pass

    def emit(self, event: TraceEvent) -> None:
        pass

    def op_start(self, *_: Any, **__: Any) -> None:
        pass

    def op_end(self, *_: Any, **__: Any) -> None:
        pass

    def op_yield(self, *_: Any, **__: Any) -> None:
        pass

    def annotate(self, *_: Any, **__: Any) -> None:
        pass

    def llm_usage(self, *_: Any, **__: Any) -> None:
        pass

    def media_ref(self, *_: Any, **__: Any) -> None:
        pass

    @contextmanager
    def group(self, name: str, **metadata: Any) -> Iterator[None]:
        yield

    def start_time_of(self, *_: Any) -> float:
        return 0.0


_NULL_EMITTER = NullEmitter()
"""Singleton — returned by ``current_emitter()`` when nothing is bound.
Safe to share: NullEmitter has no per-call state."""

File: rust/operonx/src/core/tracing/emitter.rs
//! `EventEmitter` — what ops + scheduler call to record trace events.
//!
//! Mirrors Python [`operonx/core/tracing/emitter.py`](../../../../../operonx/core/tracing/emitter.py).
//! All emit methods are sync + O(1) per execution-model rule §3.7 — events
//! land on the bound pipeline's buffer immediately and the actual processor
//! / exporter work happens off the hot path inside `pipeline.flush()`.
//!
//! ## Thread-local current op (Rust analogue of Python's ContextVar)
//!
//! Python uses two `ContextVar`s (`_current_emitter_var`, `_current_op_var`)
//! that propagate across `asyncio.create_task`. The Rust runtime uses
//! `tokio::task_local!` macros for the same shape — tasks spawned with
//! `LocalKey::scope` inherit the binding; tasks spawned outside don't, which
//! mirrors Python's `asyncio.create_task` semantics.

use std::collections::HashMap;
use std::sync::Arc;
use std::time::Instant;

use chrono::Utc;
use parking_lot::Mutex;
use serde_json::Value;

use super::events::{EventCtx, EventKind, TraceEvent};
use super::pipeline::TracePipeline;

tokio::task_local! {
    /// Per-task binding of `(op_name, ctx)` for the op currently executing.
    /// Set by the scheduler around each op invocation; read by
    /// `EventEmitter::annotate` to scope user-attached metadata.
    pub static CURRENT_OP: (String, EventCtx);
}

/// Sync, fast emit surface. Owns a per-call seq counter + a
/// `(op_name, ctx) -> start_time` map for cancel-emit duration synthesis
/// (Rule 3). Clone-cheap — `Arc<Inner>`.
#[derive(Clone)]
pub struct EventEmitter {
    inner: Arc<EmitterInner>,
}

struct EmitterInner {
    pipeline: Option<Arc<TracePipeline>>,
    request_id: String,
    seq: Mutex<u64>,
    start_times: Mutex<HashMap<(String, EventCtx), Instant>>,
}

impl EventEmitter {
    /// Bind to a pipeline for the duration of one `engine.start()` call.
    pub fn new(pipeline: Arc<TracePipeline>, request_id: impl Into<String>) -> Self {
        Self {
            inner: Arc::new(EmitterInner {
                pipeline: Some(pipeline),
                request_id: request_id.into(),
                seq: Mutex::new(0),
                start_times: Mutex::new(HashMap::new()),
            }),
        }
    }

    /// A no-op emitter — every method is a cheap fast-path. Used when no
    /// tracer is wired so call sites don't need null checks.
    pub fn null() -> Self {
        Self {
            inner: Arc::new(EmitterInner {
                pipeline: None,
                request_id: String::new(),
                seq: Mutex::new(0),
                start_times: Mutex::new(HashMap::new()),
            }),
        }
    }

    /// `true` when this emitter is bound to a real pipeline. Used by callers
    /// that want to skip expensive payload assembly when nothing's listening.
    pub fn is_active(&self) -> bool {
        self.inner.pipeline.is_some()
    }

    /// Push an event onto the pipeline's buffer.
    pub fn emit(&self, event: TraceEvent) {
        if let Some(p) = &self.inner.pipeline {
            p.push(event);
        }
    }

    pub fn op_start(&self, op_name: &str, ctx: &EventCtx, inputs: Value) {
        if !self.is_active() {
            return;
        }
        self.inner
            .start_times
            .lock()
            .insert((op_name.to_string(), ctx.clone()), Instant::now());
        let event = self.build(EventKind::OpStart, Some(op_name), ctx, payload_one("inputs", inputs));
        self.emit(event);
    }

    pub fn op_end(
        &self,
        op_name: &str,
        ctx: &EventCtx,
        outputs: Value,
        status: &str,
        duration_ms: Option<f64>,
        yield_count: u32,
    ) {
        if !self.is_active() {
            return;
        }
        let key = (op_name.to_string(), ctx.clone());
        // Idempotent — second call (e.g. cancel-emit racing finally) is a no-op.
        let start = self.inner.start_times.lock().remove(&key);
        let duration = duration_ms.unwrap_or_else(|| match start {
            Some(t) => t.elapsed().as_secs_f64() * 1000.0,
            None => return_no_op(0.0),
        });
        if start.is_none() && duration_ms.is_none() {
            return;
        }
        let mut payload = serde_json::Map::new();
        payload.insert("outputs".into(), outputs);
        payload.insert("status".into(), Value::String(status.to_string()));
        payload.insert(
            "duration_ms".into(),
            serde_json::Number::from_f64(duration)
                .map(Value::Number)
                .unwrap_or(Value::Null),
        );
        payload.insert(
            "yield_count".into(),
            Value::Number(serde_json::Number::from(yield_count)),
        );
        let event = self.build(EventKind::OpEnd, Some(op_name), ctx, into_btree(payload));
        self.emit(event);
    }

    pub fn op_yield(&self, op_name: &str, ctx: &EventCtx, yielded: Value, idx: u64) {
        if !self.is_active() {
            return;
        }
        let mut payload = serde_json::Map::new();
        payload.insert("yielded".into(), yielded);
        payload.insert(
            "idx".into(),
            Value::Number(serde_json::Number::from(idx)),
        );
        let event = self.build(EventKind::OpYield, Some(op_name), ctx, into_btree(payload));
        self.emit(event);
    }

    /// Attach metadata to the currently executing op. Reads the
    /// `CURRENT_OP` task-local. Silently drops the event when called outside
    /// an op scope — mirrors Python's LookupError-as-programming-error
    /// pattern but doesn't panic the runtime.
    pub fn annotate(&self, key: &str, value: Value) {
        if !self.is_active() {
            return;
        }
        let (op_name, ctx) = match CURRENT_OP.try_with(|cur| cur.clone()) {
            Ok(v) => v,
            Err(_) => return,
        };
        let mut payload = serde_json::Map::new();
        payload.insert("key".into(), Value::String(key.to_string()));
        payload.insert("value".into(), value);
        let event = self.build(EventKind::Annotation, Some(&op_name), &ctx, into_btree(payload));
        self.emit(event);
    }

    pub fn llm_usage(
        &self,
        op_name: &str,
        ctx: &EventCtx,
        model: &str,
        prompt_tokens: u64,
        completion_tokens: u64,
        total_tokens: u64,
        cost_usd: f64,
    ) {
        if !self.is_active() {
            return;
        }
        let mut payload = serde_json::Map::new();
        payload.insert("model".into(), Value::String(model.to_string()));
        payload.insert("prompt_tokens".into(), Value::Number(prompt_tokens.into()));
        payload.insert("completion_tokens".into(), Value::Number(completion_tokens.into()));
        payload.insert("total_tokens".into(), Value::Number(total_tokens.into()));
        payload.insert(
            "cost_usd".into(),
            serde_json::Number::from_f64(cost_usd)
                .map(Value::Number)
                .unwrap_or(Value::Null),
        );
        let event = self.build(EventKind::LlmUsage, Some(op_name), ctx, into_btree(payload));
        self.emit(event);
    }

    pub fn media_ref(
        &self,
        op_name: &str,
        ctx: &EventCtx,
        handle: &str,
        mime: &str,
        size_bytes: u64,
    ) {
        if !self.is_active() {
            return;
        }
        let mut payload = serde_json::Map::new();
        payload.insert("handle".into(), Value::String(handle.to_string()));
        payload.insert("mime".into(), Value::String(mime.to_string()));
        payload.insert("size_bytes".into(), Value::Number(size_bytes.into()));
        let event = self.build(EventKind::MediaRef, Some(op_name), ctx, into_btree(payload));
        self.emit(event);
    }

    /// Open a manual group scope. Emits `GroupStart` immediately and returns
    /// a guard whose `Drop` emits `GroupEnd`. Most callers won't need this —
    /// the `GroupBy` processor synthesises boundaries from the event stream.
    pub fn group<'a>(&'a self, name: &str) -> GroupGuard<'a> {
        if self.is_active() {
            let mut payload = serde_json::Map::new();
            payload.insert("name".into(), Value::String(name.to_string()));
            let event = self.build(EventKind::GroupStart, None, &Vec::new(), into_btree(payload));
            self.emit(event);
        }
        GroupGuard {
            emitter: self,
            name: name.to_string(),
        }
    }

    /// `perf_counter`-equivalent for the op's OP_START, or `None` when
    /// unknown (cancel-before-start). Surfaces a duration_ms the scheduler
    /// can pass into `op_end` on cancel emission (Rule 3).
    pub fn start_time_of(&self, op_name: &str, ctx: &EventCtx) -> Option<Instant> {
        self.inner
            .start_times
            .lock()
            .get(&(op_name.to_string(), ctx.clone()))
            .copied()
    }

    fn build(
        &self,
        kind: EventKind,
        op_name: Option<&str>,
        ctx: &EventCtx,
        payload: std::collections::BTreeMap<String, Value>,
    ) -> TraceEvent {
        let mut s = self.inner.seq.lock();
        let seq = *s;
        *s = seq + 1;
        drop(s);
        TraceEvent {
            event_id: format!("{}-{}", self.inner.request_id, seq),
            request_id: self.inner.request_id.clone(),
            kind,
            op_name: op_name.map(String::from),
            ctx: ctx.clone(),
            timestamp: Utc::now(),
            seq,
            payload,
        }
    }
}

/// Returned by `EventEmitter::group`. Emits `GroupEnd` on drop.
pub struct GroupGuard<'a> {
    emitter: &'a EventEmitter,
    name: String,
}

impl Drop for GroupGuard<'_> {
    fn drop(&mut self) {
        if !self.emitter.is_active() {
            return;
        }
        let mut payload = serde_json::Map::new();
        payload.insert("name".into(), Value::String(self.name.clone()));
        payload.insert("status".into(), Value::String("ok".into()));
        let event = self.emitter.build(EventKind::GroupEnd, None, &Vec::new(), into_btree(payload));
        self.emitter.emit(event);
    }
}

fn payload_one(key: &str, value: Value) -> std::collections::BTreeMap<String, Value> {
    let mut m = std::collections::BTreeMap::new();
    m.insert(key.to_string(), value);
    m
}

fn into_btree(m: serde_json::Map<String, Value>) -> std::collections::BTreeMap<String, Value> {
    let mut out = std::collections::BTreeMap::new();
    for (k, v) in m {
        out.insert(k, v);
    }
    out
}

/// `unwrap_or_else` companion — returns the constant unused, used to mark
/// the branch as a no-op for clarity.
fn return_no_op<T>(t: T) -> T {
    t
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::core::tracing::pipeline::TracePipeline;

    #[test]
    fn null_emitter_is_inactive_and_silent() {
        let e = EventEmitter::null();
        assert!(!e.is_active());
        e.op_start("op", &vec!["main".into()], serde_json::json!({}));
        e.op_end("op", &vec!["main".into()], serde_json::json!({}), "ok", None, 0);
    }

    #[test]
    fn op_start_op_end_round_trip_records_duration() {
        let pipeline = Arc::new(TracePipeline::new());
        let emitter = EventEmitter::new(pipeline.clone(), "req-1");
        emitter.op_start("op", &vec!["main".into()], serde_json::json!({"x": 1}));
        std::thread::sleep(std::time::Duration::from_millis(2));
        emitter.op_end(
            "op",
            &vec!["main".into()],
            serde_json::json!({"y": 2}),
            "ok",
            None,
            0,
        );
        let events = pipeline.drain();
        assert_eq!(events.len(), 2);
        assert_eq!(events[0].kind, EventKind::OpStart);
        assert_eq!(events[1].kind, EventKind::OpEnd);
        let duration = events[1]
            .payload
            .get("duration_ms")
            .and_then(|v| v.as_f64())
            .expect("duration_ms set");
        assert!(duration >= 1.0, "duration_ms = {duration}");
    }

    #[test]
    fn op_end_idempotent_when_called_twice() {
        let pipeline = Arc::new(TracePipeline::new());
        let emitter = EventEmitter::new(pipeline.clone(), "req-2");
        emitter.op_start("op", &vec!["main".into()], serde_json::json!({}));
        emitter.op_end("op", &vec!["main".into()], serde_json::json!({}), "ok", None, 0);
        emitter.op_end("op", &vec!["main".into()], serde_json::json!({}), "ok", None, 0);
        let events = pipeline.drain();
        // 1 OpStart + 1 OpEnd; the second OpEnd was silently dropped.
        assert_eq!(events.len(), 2);
    }

    #[test]
    fn group_guard_emits_start_and_end_on_drop() {
        let pipeline = Arc::new(TracePipeline::new());
        let emitter = EventEmitter::new(pipeline.clone(), "req-3");
        {
            let _g = emitter.group("preflight");
        }
        let events = pipeline.drain();
        assert_eq!(events.len(), 2);
        assert_eq!(events[0].kind, EventKind::GroupStart);
        assert_eq!(events[1].kind, EventKind::GroupEnd);
    }

    #[tokio::test]
    async fn annotate_reads_current_op_task_local() {
        let pipeline = Arc::new(TracePipeline::new());
        let emitter = EventEmitter::new(pipeline.clone(), "req-4");
        let scope_op = ("the_op".to_string(), vec!["main".into()]);
        CURRENT_OP
            .scope(scope_op, async {
                emitter.annotate("user_id", serde_json::json!("u-1"));
            })
            .await;
        let events = pipeline.drain();
        assert_eq!(events.len(), 1);
        assert_eq!(events[0].kind, EventKind::Annotation);
        assert_eq!(events[0].op_name.as_deref(), Some("the_op"));
    }
}
